### DISCUSSION TITLE
How could I load fonts with non-latin characters more efficiently?

### DISCUSSION THREAD

**Discussion by joshli** [ocornut/imgui#3843] [Old/Archived discussions]
I'm loading fonts in my present function using the following code:

```
ImGuiIO& io = ImGui::GetIO();
ImFontConfig config;
config.OversampleH = 1;
config.OversampleV = 1;
config.GlyphExtraSpacing.x = 1.0f;
config.SizePixels = size_pixels;
config.MergeMode = true;
io.Fonts->AddFontFromFileTTF(
	"C:/Windows/Fonts/malgun.ttf", size_pixels, &config, io.Fonts->GetGlyphRangesKorean());
io.Fonts->AddFontFromFileTTF(
	"C:/Windows/Fonts/msgothic.ttc", size_pixels, &config, io.Fonts->GetGlyphRangesJapanese());
io.Fonts->AddFontFromFileTTF(
	"C:/Windows/Fonts/msyh.ttc", size_pixels, &config, io.Fonts->GetGlyphRangesChineseSimplifiedCommon());
io.Fonts->Build();
```

I found this can be quite slow and freeze the UI thread. I have tried to move this initialization to a background thread before present is called. This seems to improve things. However when `ImGui_ImplDX11_CreateFontsTexture` is called in `ImGui_ImplDX11_NewFrame`, the UI will still hang. Would anyone happen to have any suggestions on some ways I can improve this?

**Comment by rokups** [comment]
There is no better way. Best you can do is load all your fonts during program initialization where such hang is less of an intrusion. We are working on font atlas improvements that would allow on-demand rasterization of characters, but that will materialize a good while in the future.

**Comment by ocornut** [comment]
It's confusing why you say you are doing this in your "present function" since the atlas needs to be built prior to the frame beginning (first `NewFrame()` call) and therefore the "stalling" should be loading.

- Those three files you linked totals to about 35 MB, preloading the TTF files ahead and calling `AddFontFromMemory` will save a little bit.
- Generally it is not very recommended to merge large ranges. To be bluntly fair our current suggestion to load those arbitrary Asian ranges is quite subpar in the first place, but adding 3 of them is going to be worse.
- It would be better if you created a ImFontGlyphRangeBuilder and only added glyphs that you are actually using, but that's reliant on you knowing your text data ahead of time, which can be easy or difficult depending on your localization setup.

As Rokups say we are well aware that's an issue and aim to work on meaningful improvement in the next few months hopefully.

(Bonus link https://github.com/ocornut/imgui/blob/master/docs/FONTS.md)

**Reply by vmeisoft** [reply]
Can you provide an interface for me to draw text on my own?

My software aims to support all languages, but it is not feasible to load all fonts into memory. However, I can use DirectWriter to draw text without worrying about font issues.

**Reply by ocornut** [reply]
You can use the ImDrawList::AddText api to add text anywhere at lower level.

we cannot replace it presently. You cannot support every languages with dear imgui.

**Reply by vmeisoft** [reply]
I modified the code to load Chinese characters and other fonts. Although the efficiency is not high, it can also avoid loading all of them into memory. The code is as follows:
Is it possible to load only the required fonts?

static ImVector<ImWchar> ranges;
extern bool imgui_font_need_reload;
extern void imgui_font_reload();

const ImFontGlyph* ImFont::FindGlyph(ImWchar c) const
{
    if (c >= (size_t)IndexLookup.Size)
    {
        imgui_add_char(c);
        return FallbackGlyph;
    }
    const ImWchar i = IndexLookup.Data[c];
    if (i == (ImWchar)-1)
    {
        imgui_add_char(c);
        return FallbackGlyph;
    }
    return &Glyphs.Data[i];
}

void imgui_font_reload()
{
    ImGuiIO& io = ImGui::GetIO();
    io.Fonts->Clear();
    ranges.clear();
    builder.BuildRanges(&ranges);
    ImFont* newFont = io.Fonts->AddFontFromFileTTF(FontPathList[imgui_font_index].Path, imgui_font_size, NULL, ranges.Data);
    io.Fonts->Build();
    io.FontDefault = newFont;
}

void imgui_add_char(wchar_t c)
{
    if (!builder.GetBit(c))
    {
        builder.AddChar(c);
        Chars += c;
        SaveStringValue(CHARS_NAME, Chars);
        imgui_font_need_reload = true;
    }
}

void ImGui_ImplDX11_NewFrame()
{
    ImGui_ImplDX11_Data* bd = ImGui_ImplDX11_GetBackendData();
    IM_ASSERT(bd != nullptr && "Did you call ImGui_ImplDX11_Init()?");

    if (!bd->pFontSampler)
    {
        imgui_font_need_reload = false;
        ImGui_ImplDX11_CreateDeviceObjects();
    }
    else if (imgui_font_need_reload)
    {
        imgui_font_need_reload = false;
        imgui_font_reload();
        if (bd->pFontTextureView) { bd->pFontTextureView->Release(); bd->pFontTextureView = nullptr;}
        ImGui_ImplDX11_CreateFontsTexture();
    }
}
